add env var AEON_CPU_LIST  & config attr cpu_list which control cpu affinity

sfraczek opened the pull request:
Now thread affinity range can be specified through environment variable `AEON_CPU_LIST`.  Alternatively use `cpu_list` argument in loader config.

**Example:**
`AEON_CPU_LIST=5,10-12` 
`{cpu_list, "5,10-12"}`

the indexing starts at 0 like in *numa*. 
If unspecified, the cores are selected automatically .

Commit:
add environment variable AEON_CPU_LIST controlled affinity

Commit:
WIP review fixes from Adrian.

replaced decode thread count with cpu list

Commit:
review fixes

added config variable cpu_list

moved automatic core number selection to loader.cpp

updated documentation

small refactorings

commented out unreachable input validation code in cpu_list parsing

Commit:
review fix, remove comment

sfraczek commented:
> If this should never happen, please remove this commented code.

ok done

arogowie-intel commented on src/loader.cpp:
```diff
@@ -179,16 +180,44 @@ void loader_local::initialize(const json& config_json)
+        thread_affinity_map.resize(nthreads);
+        std::iota(thread_affinity_map.begin(), thread_affinity_map.end(), 0);
+    }
+    unsigned int threads_num = !thread_affinity_map.empty() ? thread_affinity_map.size()
```
I think here you don't need to check for _non emptyness_ of `thread_affinity_map`. In above code you fill it in case there is no `cpu_list` nor `AEON_CPU_LIST` variables set. Thus I think the following line would suffice:
```suggestion
    unsigned int threads_num = thread_affinity_map.size();
```
Moreover If `thread_affinity_map.empty()` would be true, why setting `threads_num` to `std::thread::hardware_concurrency()` ? I wonder whether this would be correct

sfraczek replied:
Thanks @arogowie-intel . You are right. There is something wrong. I will look closer at this because I'm not sure what @AdrShv meant here for threads_num variable used below in line 215 here https://github.com/NervanaSystems/aeon/pull/9/files/b70b736b96c97fc5b0f152bc06667dbf8475d99c#diff-50f539ac61d3b1b208f9d379fdea5153R215. It has different calculation than https://github.com/NervanaSystems/aeon/pull/9/files/b70b736b96c97fc5b0f152bc06667dbf8475d99c#diff-50f539ac61d3b1b208f9d379fdea5153R204 so I think it should still be a different number but I got it wrong.

arogowie-intel commented on src/loader.cpp:
```diff
@@ -179,16 +180,44 @@ void loader_local::initialize(const json& config_json)
     m_decoder = make_shared<batch_decoder>(m_batch_iterator,
                                            decode_size,
-                                           lcfg.decode_thread_count,
+                                           thread_affinity_map,
```
Why not moving this `thread_affinity_map` instead of copying?

sfraczek replied:
I was hoping this would come out in a review so we can discuss a suggestion of how to optimize this. I'm not sure if we can leave it to the compiler to do a better job or explicitly implement std;:move or const reference passing or other ideas.

sfraczek replied:
ok I added some move semantics

AdrShv replied:
IMHO, const reference is the best choice, because we are not going to change that vector

AdrShv replied:
so we don't  have to use move semantics there

sfraczek replied:
What do you mean?

sfraczek replied:
thread_pool now makes a copy of it, I'm not sure if we can have thread_pool hold a const reference instead because it might get destroyed when leaves the scope of `void loader_local::initialize(const json& config_json)`.
So now I think the std::move is the best.

sfraczek replied:
using r-value references or perfect forwarding made things ugly so I moved to const reference with single copy because its not a critical path or a heavy copying operation

arogowie-intel commented on src/thread_pool.hpp:
```diff
@@ -83,24 +83,10 @@ template <typename T, void (T::*process_func)(int index)>
 {
 public:
-    thread_pool(int thread_count)
+    thread_pool(std::vector<int> thread_affinity_map)
```
It think it would be better to accept this as a reference or a r-value reference and from such initialize member `m_thread_affinity_map`.

sfraczek replied:
I've read a suggestion that if I pass-by-value  an r-value using std::move it still moves instead of copying.

sfraczek replied:
the thing I did moves every time I do that instead of doing final copy. i wold have to use forwarding which makes it more messy than its worth it to optimize this when its not critical path.

arogowie-intel commented on src/thread_pool.hpp:
```diff
@@ -201,8 +186,8 @@ template <typename T, void (T::*process_func)(int index)>
 public:
-    thread_pool_queue(int thread_count)
-        : m_thread_pool(thread_count)
+    thread_pool_queue(std::vector<int> thread_affinity_map)
```
Analogous comment as above - consider accepting reference or r-value reference to avoid copying.

sfraczek replied:
answer as above

arogowie-intel commented:
Overall looks great! I like this parsing utility function :)

Commit:
use const reference for passing cpu_list

move semantics make it ugly while this is not a heavy copy anyway

arogowie-intel commented on src/loader.cpp:
```diff
@@ -179,16 +180,44 @@ void loader_local::initialize(const json& config_json)
+                     static_cast<int>(std::thread::hardware_concurrency() / m_free_threads_ratio));
+        thread_affinity_map.resize(nthreads);
+        std::iota(thread_affinity_map.begin(), thread_affinity_map.end(), 0);
+    }
```
I'd suggest to separate above two `if` conditions into separate function called sth similiar to `get_thread_affinity_map` to get more clean code.

sfraczek replied:
ok

arogowie-intel commented on src/thread_pool.hpp:
```diff
@@ -138,8 +124,7 @@ class nervana::thread_pool
 private:
-    const int                       m_max_count_of_free_threads = 2;
-    const int                       m_free_threads_ratio        = 8;
+    const std::vector<int>          m_thread_affinity_map;
```
Shouldn't this be a `const &` ? Otherwise you'll be still copying.

sfraczek replied:
I'm assuming that I need a copy here.

sfraczek replied:
I did it how Adrian told me to do it today.

Commit:
review fixes

added get_thread_affinity_map function to clean-up
fixed decode_size equation
changed constructor of thread_pool

sfraczek commented on src/util.cpp:
```diff
@@ -330,3 +330,95 @@ std::string nervana::vector2string(const std::vector<char>& v)
+    }
+    thread_affinity_map.resize(std::distance(thread_affinity_map.begin(), ip));
+
+    if (thread_affinity_map.back() > std::thread::hardware_concurrency())
```
```suggestion
    if (thread_affinity_map.back() >= std::thread::hardware_concurrency())
```

Commit:
change condition from > to >=

threads are numbered from 0 so max thread_id is `hardware_concurrency()-1`

AdrShv commented on src/batch_decoder.cpp:
```diff
@@ -32,8 +32,8 @@ batch_decoder::batch_decoder(shared_ptr<batch_iterator>                 b_itor,
-        singleton<thread_pool_queue<batch_decoder, &batch_decoder::process>>::get(thread_count);
+    m_thread_pool = singleton<thread_pool_queue<batch_decoder, &batch_decoder::process>>::get(
+        thread_affinity_map);
     m_number_elements_in = prov->get_input_count();
```
I propose to pass thread_affinity_map by value , line 26
and add std::move() here

sfraczek replied:
done

Commit:
review fixes

error msg to use hardware_concurrenty()-1,
move constructor in batch decoder
